[libc++] Implement P3168R2: Give optional range support

smallp-o-p opened the pull request:
Resolves #105430

- Implement all required pieces of P3168R2
- Leverage existing `wrap_iter` and `bounded_iter` classes to implement the `optional` regular and hardened iterator type, respectively 
- Update documentation to match

smallp-o-p committed:
Implement P2988R12: Give optional range support

smallp-o-p committed:
Misinputted columns in Cxx2cPapers.csv

H-G-Hristov commented on libcxx/docs/Status/Cxx2cPapers.csv:
```diff
@@ -66,7 +66,7 @@
 "`P2997R1 <https://wg21.link/P2997R1>`__","Removing the common reference requirement from the indirectly invocable concepts","2024-06 (St. Louis)","|Complete|","19","Implemented as a DR against C++20. (MSVC STL and libstdc++ will do the same.)"
 "`P2389R2 <https://wg21.link/P2389R2>`__","``dextents`` Index Type Parameter","2024-06 (St. Louis)","|Complete|","19",""
-"`P3168R2 <https://wg21.link/P3168R2>`__","Give ``std::optional`` Range Support","2024-06 (St. Louis)","","",""
+"`P3168R2 <https://wg21.link/P3168R2>`__","Give ``std::optional`` Range Support","2024-06 (St. Louis)","|Complete|","21",""
```
```suggestion
"`P3168R2 <https://wg21.link/P3168R2>`__","Give ``std::optional`` Range Support","2024-06 (St. Louis)","|Complete|","22",""
```
More likely

H-G-Hristov commented on libcxx/include/optional:
```diff
@@ -233,6 +246,13 @@ namespace std {
 #  include <initializer_list>
 #  include <version>
 
+#  if _LIBCPP_STD_VER >= 26
+#    include <__cstddef/ptrdiff_t.h>
+#    include <__format/range_format.h>
+#    include <__iterator/wrap_iter.h>
+#    include <__ranges/enable_view.h>
+#  endif
```
I think we do includes unconditionally mostly.

H-G-Hristov commented on libcxx/test/std/utilities/optional/iterator.pass.cpp:
```diff
@@ -0,0 +1,124 @@
+
+// template <class T> class optional::iterator;
+
+// 1: optional::iterator and optional const_iterator satisfy contiguous_iterator and random_access_iterator
+// 2. The value types and reference types for optional::iterator and optional::const_iterator are {_Tp, _Tp&} and {_Tp, const _Tp&} respectively
+// 3: The optional::begin() and optional::end() are marked noexcept.
+// 4: optionals that have a value have begin() != end(), whereas one that doesn't has begin() == end();
+// 5: The corresponding size for the following optionals is respected: has_value() == 1, !has_value() == 0
+// 6: Dereferencing an engaged optional's iterator returns the correct value.
+// 7: std::ranges::enable_view<optional<T>> == true, and std::format_kind<optional<T>> == true
+// 8: Verify that an iterator for loop counts only 1 item for an engaged optional, and 0 for an unegaged one.
+// 9: An optional with value that is reset will have a begin() == end(), then when it is reassigned a value, begin() != end(), and *begin() will contain the new value.
```
It's better to put these comments where the tests are, instead of indexing like that. It will be clearer.

H-G-Hristov commented on libcxx/test/std/utilities/optional/iterator.pass.cpp:
```diff
@@ -0,0 +1,124 @@
+#include "test_macros.h"
+#include "check_assertion.h"
+
+constexpr int test_loop(const std::optional<char> val) {
```
test_loop isn't very descriptive. Isn't that test the same as when you use ranges::size?

smallp-o-p replied:
It pretty much is, just wanted to have at least an example from the paper itself as a tests, though to be honest it's probably better to just get rid of it.

Zingam commented on libcxx/docs/ReleaseNotes/21.rst:
```diff
@@ -53,7 +53,7 @@ Implemented Papers
 - P2770R0: Stashing stashing ``iterators`` for proper flattening (`Github <https://github.com/llvm/llvm-project/issues/105250>`__)
 - P2655R3: ``common_reference_t`` of ``reference_wrapper`` Should Be a Reference Type (`Github <https://github.com/llvm/llvm-project/issues/105260>`__)
-
+- P3168R2 Give ``std::optional`` Range Support (`Github <https://github.com/llvm/llvm-project/issues/105430>`__)
 Improvements and New Features
```
Same commenter as above.
I guess fixing this will also fix the docs build.

Zingam commented:
Thanks for working on this. A nice pre-requisit to `optional<T&>`. If you decide to work on it, please let me know. I have it assigned to me but I've done minimal work yet.

smallp-o-p committed:
Address most comments

smallp-o-p committed:
Clang-Format :^)

smallp-o-p committed:
Mostly revert conditional inclusion change (for now) due to a circular inclusion in __pstl/default.h in C++17 and C++20

The circular include looks like the following:
optional -> __format/range_format.h -> __format/concepts.h -> __format/format_parse_context.h -> string_view -> algorithm -> __algorithm/pstl.h -> __pstl/backend.h -> __pstl/backends/default.h -> optional

smallp-o-p committed:
Remove mentions of implementation in Libc++ 21, should go to Libc++ 22, also remove check_assertion.h header in test due to it breaking

Zingam commented on libcxx/docs/Status/Cxx2cPapers.csv:
```diff
@@ -66,7 +66,7 @@
 "`P2997R1 <https://wg21.link/P2997R1>`__","Removing the common reference requirement from the indirectly invocable concepts","2024-06 (St. Louis)","|Complete|","19","Implemented as a DR against C++20. (MSVC STL and libstdc++ will do the same.)"
 "`P2389R2 <https://wg21.link/P2389R2>`__","``dextents`` Index Type Parameter","2024-06 (St. Louis)","|Complete|","19",""
-"`P3168R2 <https://wg21.link/P3168R2>`__","Give ``std::optional`` Range Support","2024-06 (St. Louis)","","",""
+"`P3168R2 <https://wg21.link/P3168R2>`__","Give ``std::optional`` Range Support","2024-06 (St. Louis)","|Complete|","22",""
```
https://github.com/llvm/llvm-project/commit/167c695cece8e4fa78b7e9c5fc94bae3821ade52#diff-b312e35a59a5cb780abaac0ac208061394219c78f6d301e1c17c0288b8a62cbfR38

You can update the release notes for LLVM22 now. Please mind the proper syntax.

smallp-o-p committed:
Add entry to libc++ 22 release notes

ldionne reviewed (changes requested):
Thanks a lot for the patch! This seems to check most of the boxes for what we're usually looking in a patch, which is great. I do have some comments and a few requests for changes.

smallp-o-p committed:
implicit bool conversions are evil

smallp-o-p committed:
Split test file into three tests

smallp-o-p committed:
Remove usage of remove_cvref_t for pointer and const pointer types since they weren't necessary and were messing with _Tp of const types

smallp-o-p committed:
Fix a comment, forgot a noexcept test

smallp-o-p committed:
Copy mistake

Zingam commented on libcxx/test/std/utilities/optional/optional.iterator/end.pass.cpp:
```diff
@@ -0,0 +1,73 @@
+  assert(test<int>());
+  assert(test<char>());
+  assert(test<const int>());
+  assert(test<const char>());
```
I think I like how you organized the tests now. FYI only - I am not asking for any changes: there is an alternative to iterate over multiple types a helper `types::for_each(types::integer_types(), IntegerTypesTest{});` an example usage is here: https://github.com/llvm/llvm-project/blob/9ae7490c6251ec60ab978bb818e89ca586e0c9c9/libcxx/test/std/ranges/range.factories/range.iota.view/indices.pass.cpp

frederick-vs-ja commented on libcxx/include/optional:
```diff
@@ -586,8 +614,17 @@ class _LIBCPP_DECLSPEC_EMPTY_BASES optional
 public:
   using value_type = _Tp;
+#    if _LIBCPP_STD_VER >= 26
+  using iterator       = __wrap_iter<pointer>;
```
Considering `optional<T(&)[]>` (note the missing bound) or `optional<T(&)(Args)>`, I wonder whether they should have `iterator`. Certainly, such an `iterator` can't be a contiguous iterator (and not even a real iterator in the case of `optional<T(&)(Args)>`), and our wrappers don't seem to work with them. If we decide they should have `iterator`, we probably need to introduce `optional`-specific iterators. @philnik777 @ldionne

ldionne replied:
@philnik777 The improvement I see by having an optional-specific iterator is that we can't mistake it for a `std::vector::iterator`, for example. There might also be optimization opportunities by knowing that an iterator points to at most one element (basically that encodes the knowledge of a range having at most N elements where `N` is a compile-time constant). But I'm primarily interested in making the following fail:

```
std::vector<int> v = {...};
std::optional<int> opt = ...;

std::sort(v.begin(), opt.end()); // wtf?
```

If we use `__wrap_iter`, that code would compile despite being obviously wrong. A bunch of other stuff would also work, such as `v.begin() != opt.end()`, and none of it makes sense.

smallp-o-p replied:
> Finally, we'd also need to make sure that we have positive testing coverage for std::optional<int (&)[]>::pointer and std::optional<void (&)(int, char)>::pointer under libcxx/test/std. That actually makes me realize that this patch is missing such coverage even for std::optional<int> (unless I missed it) -- so it should be added. We generally test such things in a test named some/path/types.pass.cpp, and it looks like we already have one for std::optional here: libcxx/test/std/utilities/optional/optional.object/types.pass.cpp. You will want to augment it with additional tests guarded under TEST_STD_VER >= 26, a macro that can be found in test_macros.h.

There aren't any tests for `std::optional::pointer` because there isn't a requirement in the standard about it, just the `iterator` type. I was just following what other containers were doing with `__wrap_iter`. It's also why they're private, I realize they should've been named as an internal type, which is a miss on my part.

frederick-vs-ja commented on libcxx/test/libcxx/transitive_includes/cxx26.csv:
```diff
@@ -2,6 +2,7 @@ algorithm cctype
 algorithm climits
 algorithm compare
 algorithm cstdint
+algorithm cstdio
```
We need `range_format` and `format_kind` here. But `<__format/range_format.h>` drags in `<__format/concepts.h>`, which in turn drags in `<string_view>` and several unnecessary stuffs.

smallp-o-p committed:
Add hardening with bounded_iter

smallp-o-p committed:
Fix misplaced parens

smallp-o-p committed:
Remove some transitive includes that were removed with #150583

smallp-o-p committed:
Remove include guard since it's no longer necessary

ldionne reviewed (commented):
This is looking very good!

I would be okay with moving forward with this patch after my small comments have been addressed, but I would like a commitment to introduce an iterator that is specific to `optional` in the near future (in particular before the LLVM 22 release). If we don't have an iterator that is specific to `optional`, we can't ship this in a release because this is going to be ABI once we ship it.

smallp-o-p committed:
Address comments

smallp-o-p committed:
Forgot to rename a variable

ldionne commented:
> I can get started with something like an `upper_bound_iterator` or something of the sort which could be useful for both `inplace_vector` and `optional` once this patch gets merged in.

This is neat, I like that you identified that this would also be useful for `inplace_vector`! I suggest the name is going to be a bit difficult to nail down but let's start with something and then talk about it on the review.

> Edit: It would most likely be a two-part patch: One being a `type_aware_iterator<class _Iter, class _Container>` that wraps existing iterators, but solves the first gripe about contiguous iterators being interchangeable, and finally a `upper_bound_iter<class _Iter, size_t size>`, but I'd have to think a little more about it.

Alternatively, we could simply pass the `_Container` directly to `__upper_bound_iterator` -- that would avoid creating too many layers of wrapping which wouldn't be great for compile-times and also might complicate some optimizations we might want to apply.

smallp-o-p committed:
Use std::as_const pattern for all tests

smallp-o-p committed:
Rename pointer type  and unguard it

smallp-o-p committed:
Disallow T(&)[] and T&(Args...) from having iterator types as pre-caution

smallp-o-p committed:
Forgot an #include

smallp-o-p committed:
Retrigger Build

smallp-o-p committed:
is_unbounded_array is >= C++20

smallp-o-p committed:
Bring in __pointer types too, there is probably a better way to do this--this is to avoid cluttering up error messages with __iter_aliases substitution failures when an invalid type for optional is passed

smallp-o-p committed:
Remove unnecessary public:

philnik777 commented on libcxx/include/optional:
```diff
@@ -579,16 +607,49 @@ struct __is_std_optional : false_type {};
+template <class _Tp>
+struct __optional_iterator_aliases<
+    _Tp,
+    __enable_if_t<!(is_reference<_Tp>::value && (is_function<__libcpp_remove_reference_t<_Tp>>::value ||
+                                                 is_unbounded_array<__libcpp_remove_reference_t<_Tp>>::value))> > {
```
IMO we should tackle it then. It's not clear to me that we even want to allow these kinds of instantiations, and I'm very much not a fan of working around problems that don't even exist yet.

smallp-o-p replied:
I'm not against it, but I'll ask @ldionne and @frederick-vs-ja what they think about cutting this out for a later patch.

ldionne reviewed (approved):
The patch LGTM as-is or with the `__optional_iterator_aliases` removed per @philnik777 's comment. Either way looks fine to me.

What I care strongly about is that we then tackle the iterator type confusion problem.

Thanks a lot for the patch, this is a great (first?) contribution.

Zingam commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -36,6 +36,9 @@ What's New in Libc++ 22.0.0?
 
 Implemented Papers
+
+- P3168R2: Give ``std::optional`` Range Support (`Github <https://github.com/llvm/llvm-project/issues/105430>`__)
```
You should move this line bellow P2321. This breaks the syntax.

smallp-o-p committed:
Don't worry about references

smallp-o-p committed:
Update documentation

frederick-vs-ja reviewed (commented):
> - Leverage existing `wrap_iter` class to implement the `optional` iterator type. Hardening has not been added, do we want that for `optional`?

Looks like this question has been resolved. @smallp-o-p Would you mind to update this bullet in PR description, so that it will be greater to use the description in the squashed commit message as-is?
